pkg/web/client.go

```diff
@@ -0,0 +1,239 @@
+	CogPredictCodeStrip string `json:"COG_PREDICT_CODE_STRIP"`
+	CogTrainCodeStrip   string `json:"COG_TRAIN_CODE_STRIP"`
+	CogPyEnvPath        string `json:"COG_PYENV_PATH"`
+	CogEagerImports     string `json:"COG_EAGER_IMPORTS"`
+	R8CudaVersion       string `json:"R8_CUDA_VERSION"`
+	R8CudnnVersion      string `json:"R8_CUDNN_VERSION"`
+	R8PythonVersion     string `json:"R8_PYTHON_VERSION"`
+	R8TorchVersion      string `json:"R8_TORCH_VERSION"`
+}
+
+type RuntimeConfig struct {
+	Weights []File `json:"weights"`
+	Files   []File `json:"files"`
+	Env     Env    `json:"env"`
+}
+
+type Version struct {
+	Annotations   map[string]string      `json:"annotations"`
+	CogConfig     config.Config          `json:"cog_config"`
+	CogVersion    string                 `json:"cog_version"`
+	OpenAPIScheme map[string]interface{} `json:"openapi_schema"`
+	RuntimeConfig RuntimeConfig          `json:"runtime_config"`
+	Virtual       bool                   `json:"virtual"`
+}
+
+func NewClient(command command.Command, client *http.Client) *Client {
+	return &Client{
+		command: command,
+		client:  client,
+	}
+}
+
+func (c *Client) PostNewVersion(ctx context.Context, image string, weights []File, files []File) error {
+	userInfo, err := c.command.LoadUserInformation(global.ReplicateRegistryHost)
+	if err != nil {
+		return err
+	}
+
+	version, err := c.versionFromManifest(image, weights, files)
+	if err != nil {
+		return err
+	}
+
+	jsonData, err := json.Marshal(version)
+	if err != nil {
+		return err
+	}
+
+	versionUrl := newVersionURL(userInfo.Username, image)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, versionUrl.String(), bytes.NewReader(jsonData))
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated {
+		return errors.New("Bad response from new version endpoint: " + strconv.Itoa(resp.StatusCode))
+	}
+
+	return nil
+}
+
+func (c *Client) versionFromManifest(image string, weights []File, files []File) (*Version, error) {
+	manifest, err := c.command.Inspect(image)
+	if err != nil {
+		return nil, err
+	}
+
+	var cogConfig config.Config
+	err = json.Unmarshal([]byte(manifest.Config.Labels[command.CogConfigLabelKey]), &cogConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	var openAPIScheme map[string]interface{}
+	err = json.Unmarshal([]byte(manifest.Config.Labels[command.CogOpenAPISchemaLabelKey]), &cogConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	predictCode, err := stripCodeFromStub(cogConfig, true)
+	if err != nil {
+		return nil, err
+	}
+	trainCode, err := stripCodeFromStub(cogConfig, false)
+	if err != nil {
+		return nil, err
+	}
+
+	var cogGPU int
+	if cogConfig.Build.GPU {
+		cogGPU = 1
+	}
+
+	torchVersion := ""
+	cudaVersion := ""
+	cudnnVersion := ""
+	pythonVersion := ""
+	pythonPath := ""
+	for _, env := range manifest.Config.Env {
+		envComponents := strings.Split(env, "=")
```

Comment:
ooh! strings.Cut is nice for this 

https://pkg.go.dev/strings#Cut

Reply:
Resolved in [0517e96](https://github.com/replicate/cog/pull/2139/commits/0517e96f1a23dd4bdb1cf946dce80d81702bc663)